### ISSUE TITLE
Request by Service IP is two or three times slower than by pod IP

### ISSUE THREAD

**Issue by 372046933** [kubernetes/kubernetes#87277]
**What happened**:
I have a deployed a service with two pods. 
```
ab -c 16 -n 2000 -p /PATH/TO/DATA -T 'application/json' ttp://SERVICE_IP:SERVICE_PORT/xxx
```
and 
```
ab -c 16 -n 2000 -p /PATH/TO/DATA -T 'application/json' http://POD_IP:POD_PORT/xxx
```
are test cases. Access via service IP takes about 2~3 seconds per request, while pod Ip takes about 500 ms.

**What you expected to happen**:
Since Service is backed by two pods. It should have better response time

**How to reproduce it (as minimally and precisely as possible)**:
The service is TensorFlow serving. I used the image `tensorflow/serving:2.0.0-gpu` from the DockerHub.
And each pod has requested a GPU. Batching is enabled and have timeout 100 ms.
**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`): both client and server are 1.14.8
- Cloud provider or hardware configuration: Bare metal
- OS (e.g: `cat /etc/os-release`): CentOS 7
- Kernel (e.g. `uname -a`): 3.10.0-1062.9.1
- Install tools:
- Network plugin and version (if this is a network-related bug): Flannel, host-gateway
- Others:

**Comment by athenabot** [comment]
/sig network

These SIGs are my best guesses for this issue. Please comment `/remove-sig <name>` if I am incorrect about one.

🤖 I am a bot run by vllry. 👩‍🔬

**Comment by athenabot** [comment]
/triage unresolved

Comment `/remove-triage unresolved` when the issue is assessed and confirmed.

🤖 I am a bot run by vllry. 👩‍🔬

**Comment by rikatz** [comment]
@372046933 what backend type of kube-proxy are you using? iptables or IPVS?

**Comment by 372046933** [comment]
@rikatz It's iptables. And all nodes are in the same LAN. RTT between nodes is around 50~60us. Network plugin is flannel

**Comment by rikatz** [comment]
Hum, ok. Just to start eliminating some possible cause, can you please try running kube-proxy with IPVS mode instead of iptables, or is this a production environment that you can't change.

The idea here is to eliminate some condition caused by IPtables from kubelet x kube-proxy x flannel :)

Thank you

**Comment by 372046933** [comment]
@rikatz  I have tried putting the POD's IP in Nginx. Thus no iptables is required. I also tried replacing Nginx with F5. Problem exists anyway.

**Comment by 372046933** [comment]
@rikatz By the way, the model under serving is a seq2seq model. The latency problem cannot re reproduced on other models like resnet50. Maybe the model has inherent problem on scaling?

**Comment by rikatz** [comment]
@372046933 Sorry but I cannot figure out how putting the Pod IP in something outside the cluster could be something. Can you explain me.

Just let me clarify how kube-proxy works:

When you create a ClusterIP, each node running kube-proxy will map that ClusterIP with the Pods related to them. So each node 'becames' a Load Balancer only for the cluster.

As an example, imagine you have a ClusterIP of 10.10.10.10 pointing to Pods with the IPs 10.20.20.20 (running in NodeA) and 10.20.30.20 (running in nodeB) and a client with the IP 10.20.20.21 (running in NodeA).

When you call from the client to the ServiceIP, the local kube-proxy (running in the same node of the client) containing the map of ClusterIP/Pods will direct the connection to any of the Pods. So who makes the balance and redirection of the traffic is the local kube-proxy to the Pods.

So the flow would be client -> ClusterIP (in the same node of the client) -> Pod IP (running in any of the nodes of the cluster).

So what makes this job is the local IPTables or IPVS (depending on how you configure kube-proxy). This way, this might be something related to how kube-proxy redirects this traffic.

If you can make the same test replacing the proxy-mode from kube-proxy from IPTables to IPVS we can check if this is something related to IPTables or other thing in your cluster.

Tks

**Comment by 372046933** [comment]
Thanks for your kindly help. To exclude iptable problems. We have put an Nginx in the cluster, and manually  write the pod's IP in Nginx's conf. So there is not iptable required. Right? We also 
tried using JMeter and even writing client HTTP code to redo the benchmark. The result is confusing. Both JMeter and `ab` have significant delays. While our own code don't. I will continue to report the latest results here.

**Comment by rikatz** [comment]
Actually I don't think this is right. If the Pod is inside the Kubernetes Cluster there are some other things than only putting the IP into Nginx Conf to make it work. 

Like, as you're using Flannel the CNI is responsible to estabilish the tunnels and make the environment know what's the right route to the IP. Flannel uses IPTables also, so you're not removing it from the scenario anyway :)

**Comment by 372046933** [comment]
@rikatz In our environment, the backend of Flannel is `host-gw`, and all nodes are under same L2 network. Even if I turn all `iptables` off, containers can communicate with each other using the host's route table

**Comment by rikatz** [comment]
@372046933 can you please take a look into this [comment](https://github.com/kubernetes/kubernetes/issues/86507#issuecomment-595227060 ) and check if it helps?

**Comment by 372046933** [comment]
We do not have `flannel0` interface in our environment. Each `docker0` acquires  IP from config which is written by flanneld

**Comment by rikatz** [comment]
@372046933 indeed. I'll create a lab here with kubeadm + flannel in hostgw mode and check some scenarios

**Comment by 372046933** [comment]
Great. Better with a model serving like seq2seq with dynamic RNN

**Comment by rikatz** [comment]
@372046933 trying not to be annoying (sorry about asking this much) but what's your network card model? There's a similar issue here (https://github.com/kubernetes/kubernetes/issues/88986) which is also baremetal and might be suffering from the same problem.

As a test, is it possible to you to upgrade your kernel to another version, or maybe use another Operating System?

There's some discussion aroung about the tx/rx offload (as I've commented) and to be honest, I've faced also this problem in a different scenario (not Kubernetes) but with a bug in the kernel.

Anyway I'm installing a CentOS 7.7 with the kernel 3.10.0-1062 to see if this behavior also happens here. 

Thank you
